arangod/Replication2/Streams/LogMultiplexer.h

```diff
@@ -0,0 +1,85 @@
+#pragma once
+#include <utility>
+
+#include <Futures/Future.h>
+
+#include <Replication2/ReplicatedLog/ILogParticipant.h>
+#include <Replication2/ReplicatedLog/LogCommon.h>
+#include <Replication2/ReplicatedLog/types.h>
+
+#include <Replication2/Streams/StreamSpecification.h>
+#include <Replication2/Streams/Streams.h>
+
+namespace arangodb::replication2::replicated_log {
+class LogFollower;
+class LogLeader;
+}  // namespace arangodb::replication2::replicated_log
+
+namespace arangodb::replication2::streams {
+
+/**
+ * Common stream dispatcher class for Multiplexer and Demultiplexer. You can
+ * obtain a stream given its id using getStreamById. Alternatively, you can
+ * static_cast the a pointer to StreamBase<Descriptor> for the given stream.
+ * @tparam Self
+ * @tparam Spec
+ * @tparam StreamType
+ */
+template <typename Self, typename Spec, template <typename> typename StreamType>
+struct LogMultiplexerStreamDispatcher : std::enable_shared_from_this<Self>,
+                                        StreamDispatcherBase<Spec, StreamType> {
+  template <StreamId Id, typename Descriptor = stream_descriptor_by_id_t<Id, Spec>>
+  auto getStreamBaseById()
+      -> std::shared_ptr<StreamGenericBase<Descriptor, StreamType>> {
+    return getStreamByDescriptor<Descriptor>();
+  }
+
+  template <StreamId Id, typename Descriptor = stream_descriptor_by_id_t<Id, Spec>,
+            typename Type = stream_descriptor_type_t<Descriptor>>
+  auto getStreamById() -> std::shared_ptr<StreamType<Type>> {
+    return getStreamByDescriptor<Descriptor>();
+  }
```

Comment:
It is a bit strange the these functions are named `ById`, but the result actually depends on the `Descriptor`. By default the descriptor is based on the Id, but if the descriptor is passed explicitly, then the `ById` part is confusing/misleading. I would either strip the Descriptor parameter or rename the function.